exchanges/ftx/ftx_websocket.go

```diff
@@ -0,0 +1,485 @@
+				Pair:         p,
+				AssetType:    a,
+			}
+		case wsOrderbook:
+			var resultData WsOrderbookDataStore
+			err = json.Unmarshal(respRaw, &resultData)
+			if err != nil {
+				return err
+			}
+			if len(resultData.OBData.Asks) == 0 && len(resultData.OBData.Bids) == 0 {
+				return nil
+			}
+			err = f.WsProcessUpdateOB(&resultData.OBData, p, a)
+			if err != nil {
+				f.wsResubToOB(p)
+				return err
+			}
+		case wsTrades:
+			var resultData WsTradeDataStore
+			err = json.Unmarshal(respRaw, &resultData)
+			if err != nil {
+				return err
+			}
+			for z := range resultData.TradeData {
+				var oSide order.Side
+				oSide, err = order.StringToOrderSide(resultData.TradeData[z].Side)
+				if err != nil {
+					f.Websocket.DataHandler <- order.ClassificationError{
+						Exchange: f.Name,
+						Err:      err,
+					}
+				}
+				f.Websocket.DataHandler <- wshandler.TradeData{
+					Timestamp:    resultData.TradeData[z].Time,
+					CurrencyPair: p,
+					AssetType:    a,
+					Exchange:     f.Name,
+					Price:        resultData.TradeData[z].Price,
+					Amount:       resultData.TradeData[z].Size,
+					Side:         oSide,
+				}
+			}
+		case wsOrders:
+			var resultData WsOrderDataStore
+			err = json.Unmarshal(respRaw, &resultData)
+			if err != nil {
+				return err
+			}
+			pair := currency.NewPairFromString(resultData.OrderData.Market)
+			var assetType asset.Item
+			assetType, err = f.GetPairAssetType(pair)
+			if err != nil {
+				return err
+			}
+			var oSide order.Side
+			oSide, err = order.StringToOrderSide(resultData.OrderData.Side)
+			if err != nil {
+				f.Websocket.DataHandler <- order.ClassificationError{
+					Exchange: f.Name,
+					Err:      err,
+				}
+			}
+			var resp order.Detail
+			resp.Side = oSide
+			resp.Amount = resultData.OrderData.Size
+			resp.AssetType = assetType
+			resp.ClientOrderID = resultData.OrderData.ClientID
+			resp.Exchange = f.Name
+			resp.ExecutedAmount = resultData.OrderData.FilledSize
+			resp.ID = strconv.FormatInt(resultData.OrderData.ID, 10)
+			resp.Pair = pair
+			resp.RemainingAmount = resultData.OrderData.Size - resultData.OrderData.FilledSize
+			var orderVars OrderVars
+			orderVars, err = f.compatibleOrderVars(resultData.OrderData.Side,
+				resultData.OrderData.Status,
+				resultData.OrderData.OrderType,
+				resultData.OrderData.FilledSize,
+				resultData.OrderData.Size,
+				resultData.OrderData.AvgFillPrice)
+			if err != nil {
+				return err
+			}
+			resp.Status = orderVars.Status
+			resp.Side = orderVars.Side
+			resp.Type = orderVars.OrderType
+			resp.Fee = orderVars.Fee
+			f.Websocket.DataHandler <- resp
+		case wsFills:
+			var resultData WsFillsDataStore
+			err = json.Unmarshal(respRaw, &resultData)
+			if err != nil {
+				return err
+			}
+			f.Websocket.DataHandler <- resultData.FillsData
+		default:
+			f.Websocket.DataHandler <- wshandler.UnhandledMessageWarning{Message: f.Name + wshandler.UnhandledMessage + string(respRaw)}
```

Comment:
Got a few unhandled messages:
[WARN] | 01/06/2020 15:10:23 |FTX - Unhandled websocket message: {"type": "error", "code": 400, "msg": "Not logged in"}
[WARN] | 01/06/2020 15:10:23 |FTX - Unhandled websocket message: {"type": "error", "code": 404, "msg": "No such market: BTC-MOVE"}

So make sure we handle errors as they come through.